### PR TITLE
Add event tracing and ETDumps to executor_runner

### PR DESCRIPTION
**Pull request by benkli01**
- Enabled via EXECUTORCH_ENABLE_EVENT_TRACER
- Add flag 'etdump_path' to specify the file path for the ETDump file
- Add flag 'num_executions' for number of iterations to run
- Create and pass event tracer 'ETDumpGen'
- Save ETDump to disk
- Update docs to reflect the changes

### CODE REVIEW

**Comment by benkli01** [comment]
The failing test looks unrelated to my changes.

File: docs/source/native-delegates-executorch-xnnpack-delegate.md [hunk 1]
```diff
@@ -74,7 +74,7 @@ Since weight packing creates an extra copy of the weights inside XNNPACK, We fre
...
 
 #### **Profiling**
-We have enabled basic profiling for XNNPACK delegate that can be enabled with the following compiler flag `-DENABLE_XNNPACK_PROFILING`. With ExecuTorch's SDK integration, you can also now use the SDK tools to profile the model. You can follow the steps in [Using the ExecuTorch SDK to Profile a Model](./tutorials/sdk-integration-tutorial) on how to profile ExecuTorch models and use SDK's Inspector API to view XNNPACK's internal profiling information.
+We have enabled basic profiling for XNNPACK delegate that can be enabled with the following compiler flag `-DEXECUTORCH_ENABLE_EVENT_TRACER`. With ExecuTorch's SDK integration, you can also now use the SDK tools to profile the model. You can follow the steps in [Using the ExecuTorch SDK to Profile a Model](./tutorials/sdk-integration-tutorial) on how to profile ExecuTorch models and use SDK's Inspector API to view XNNPACK's internal profiling information. An example implementation is available in the `xnn_executor_runner` (see [tutorial here](tutorial-xnnpack-delegate-lowering.md#profiling)).
```
Comment (digantdesai) [inline docs/source/native-delegates-executorch-xnnpack-delegate.md:77]: `ENABLE_XNNPACK_PROFILING` is still available for more detailed profiling like operator timing. We should reflect that.

File: docs/source/tutorial-xnnpack-delegate-lowering.md [hunk 2]
```diff
@@ -171,3 +171,6 @@ Now you should be able to find the executable built at `./cmake-out/backends/xnn
 
 ## Building and Linking with the XNNPACK Backend
 You can build the XNNPACK backend [CMake target](https://github.com/pytorch/executorch/blob/main/backends/xnnpack/CMakeLists.txt#L83), and link it with your application binary such as an Android or iOS application. For more information on this you may take a look at this [resource](demo-apps-android.md) next.
+
+## Profiling
+To enable profiling in the `xnn_executor_runner` pass the flags `-DEXECUTORCH_ENABLE_EVENT_TRACER=ON` and `-DEXECUTORCH_BUILD_SDK=ON` to the build command. This will enable ETDump generation when running the inference and enables command line flags for profiling (see `xnn_executor_runner --help` for details).
```
Comment (digantdesai) [inline docs/source/tutorial-xnnpack-delegate-lowering.md:176]: same here

File: backends/xnnpack/CMakeLists.txt [hunk 3]
```diff
@@ -113,8 +114,13 @@ if(NOT CMAKE_TOOLCHAIN_FILE MATCHES ".*(iOS|ios\.toolchain)\.cmake$")
   #
   list(TRANSFORM _xnn_executor_runner__srcs PREPEND "${EXECUTORCH_ROOT}/")
   add_executable(xnn_executor_runner ${_xnn_executor_runner__srcs})
+
+  if(EXECUTORCH_ENABLE_EVENT_TRACER)
+    list(APPEND xnn_executor_runner_libs etdump)
```
Comment (digantdesai) [inline backends/xnnpack/CMakeLists.txt:119]: I guess this never worked for OSS. Thanks. We should also consider adding `ENABLE_XNNPACK_PROFILING` CMake flag here.

File: examples/portable/executor_runner/executor_runner.cpp [hunk 4]
```diff
@@ -29,13 +30,23 @@
...
 DEFINE_string(
     model_path,
     "model.pte",
     "Model serialized in flatbuffer format.");
+#ifdef ET_EVENT_TRACER_ENABLED
+DEFINE_string(
+    etdump_path,
+    "model.etdump",
+    "If ETDump generation is enabled an ETDump will be written out to this path.");
+DEFINE_uint32(num_executions, 10, "Number of times to run the model.");
```
Comment (digantdesai) [inline examples/portable/executor_runner/executor_runner.cpp:48]: Do you want to make this available more generally outside EVENT_TRACER with default value = 1?

**Review by digantdesai — APPROVED** [review]
Looks good to me. Left some comments. I assume you tried running this with both portable and XNNPACK, with and without this flag and it works OK?

Do you want to add a test with this flag enabled just so we can catch any build issues in this part of the logic?

File: backends/xnnpack/CMakeLists.txt (hunk 3 above)
Comment (benkli01) [reply backends/xnnpack/CMakeLists.txt:119]: The flag `ENABLE_XNNPACK_PROFILING` is not used by the executor_runner to enable etdumps, so I thought it should be fine to just use the `EXECUTORCH_ENABLE_EVENT_TRACER`.

File: docs/source/native-delegates-executorch-xnnpack-delegate.md (hunk 1 above)
Comment (benkli01) [reply docs/source/native-delegates-executorch-xnnpack-delegate.md]: Good point. I updated the docs. 👍

**Comment by benkli01** [comment]
Thanks for the review @digantdesai ! I pushed a patch with some changes.

> I assume you tried running this with both portable and XNNPACK, with and without this flag and it works OK?

Yes, I didn't see any problems.

> Do you want to add a test with this flag enabled just so we can catch any build issues in this part of the logic?

Sure. Where would you add this test? I haven't found an obvious place.

File: backends/xnnpack/CMakeLists.txt (hunk 3 above)
Comment (digantdesai) [reply backends/xnnpack/CMakeLists.txt:119]: Yeah not an issue with your PR.

**Review by digantdesai — APPROVED** [review]

**Comment by GregoryComer** [comment]
Hey @benkli01, I ended up needing to back out this change due to it breaking the test-llama-runner-qnn-linux job (see https://github.com/pytorch/executorch/actions/runs/10398365231/job/28795542762). I apologize for the inconvenience, it didn't get caught on the PR. I'm happy to help re-land this, as this is a change that we want to keep. Thanks.

**Comment by digantdesai** [comment]
cc @cccclai

**Comment by benkli01** [comment]
Re-opened with https://github.com/pytorch/executorch/pull/5027 as this change was reverted.

**Pull request merged**